### ISSUE TITLE
TypeError: Argument 'None' of type '<class 'NoneType'>' is not a valid JAX type

### ISSUE THREAD

**Issue by bdatko** [pyro-ppl/numpyro#1078]
# Issue Description

I think I ran into a bug when following the [getting started 8 schools tutorial](http://num.pyro.ai/en/stable/getting_started.html#a-simple-example-8-schools) with following error:
`TypeError: Argument 'None' of type '<class 'NoneType'>' is not a valid JAX type`

I am seeing this error with Jax 0.2.16 and JaxLib 0.1.67 but not Jax 0.2.10 and JaxLib 0.1.62

## Environment
OS:
```bash
DISTRIB_ID=ManjaroLinux
DISTRIB_RELEASE=21.0.7
DISTRIB_CODENAME=Ornara
DISTRIB_DESCRIPTION="Manjaro Linux"
```
python:
```bash
Python 3.8.10
```
NumPyro version:
```bash
python -c "import numpyro; print(numpyro.__version__)"
0.6.0
```
Jax version:
```bash
python -c "import jax; print(jax.__version__)"
0.2.16
```

pip freeze:
```bash
absl-py @ file:///tmp/build/80754af9/absl-py_1623867230185/work
appdirs==1.4.4
argcomplete==1.12.3
argon2-cffi @ file:///tmp/build/80754af9/argon2-cffi_1613037097816/work
arviz @ file:///tmp/build/80754af9/arviz_1614019183254/work
async-generator @ file:///home/ktietz/src/ci/async_generator_1611927993394/work
attrs @ file:///tmp/build/80754af9/attrs_1620827162558/work
backcall @ file:///home/ktietz/src/ci/backcall_1611930011877/work
black==21.6b0
bleach @ file:///tmp/build/80754af9/bleach_1612211392645/work
certifi==2021.5.30
cffi @ file:///tmp/build/80754af9/cffi_1613246945912/work
cftime @ file:///tmp/build/80754af9/cftime_1621828183578/work
click==8.0.1
colorama==0.4.4
commitizen==2.17.11
cycler==0.10.0
decli==0.5.2
decorator @ file:///tmp/build/80754af9/decorator_1621259047763/work
defusedxml @ file:///tmp/build/80754af9/defusedxml_1615228127516/work
entrypoints==0.3
fire @ file:///home/conda/feedstock_root/build_artifacts/fire_1611345183397/work
flatbuffers @ file:///tmp/build/80754af9/python-flatbuffers_1614345733764/work
importlib-metadata @ file:///tmp/build/80754af9/importlib-metadata_1617874469820/work
ipykernel @ file:///tmp/build/80754af9/ipykernel_1596207638929/work/dist/ipykernel-5.3.4-py3-none-any.whl
ipython @ file:///tmp/build/80754af9/ipython_1598883821407/work
ipython-genutils @ file:///tmp/build/80754af9/ipython_genutils_1606773439826/work
ipywidgets @ file:///tmp/build/80754af9/ipywidgets_1610481889018/work
jax @ file:///home/conda/feedstock_root/build_artifacts/jax_1624496021534/work
jaxlib @ file:///home/conda/feedstock_root/build_artifacts/jaxlib_1621314943327/work/jaxlib-0.1.67-cp38-none-manylinux2010_x86_64.whl
jedi @ file:///tmp/build/80754af9/jedi_1611333125342/work
Jinja2 @ file:///tmp/build/80754af9/jinja2_1621238361758/work
jsonschema @ file:///tmp/build/80754af9/jsonschema_1602607155483/work
jupyter==1.0.0
jupyter-client @ file:///tmp/build/80754af9/jupyter_client_1616770841739/work
jupyter-console @ file:///tmp/build/80754af9/jupyter_console_1616615302928/work
jupyter-core @ file:///tmp/build/80754af9/jupyter_core_1612213311222/work
jupyterlab-pygments @ file:///tmp/build/80754af9/jupyterlab_pygments_1601490720602/work
jupyterlab-widgets @ file:///tmp/build/80754af9/jupyterlab_widgets_1609884341231/work
jupytext @ file:///home/conda/feedstock_root/build_artifacts/jupytext_1623785540978/work
kiwisolver @ file:///tmp/build/80754af9/kiwisolver_1612282420641/work
markdown-it-py @ file:///home/conda/feedstock_root/build_artifacts/markdown-it-py_1620659057741/work
MarkupSafe @ file:///tmp/build/80754af9/markupsafe_1621528148836/work
matplotlib @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-suite_1620506037462/work
mdit-py-plugins @ file:///home/conda/feedstock_root/build_artifacts/mdit-py-plugins_1620111679903/work
mistune==0.8.4
mkl-fft==1.3.0
mkl-random @ file:///tmp/build/80754af9/mkl_random_1618853849286/work
mkl-service==2.3.0
more-itertools @ file:///tmp/build/80754af9/more-itertools_1622818384463/work
mypy @ file:///tmp/build/80754af9/mypy_1624522079895/work
mypy-extensions==0.4.3
nbclient @ file:///tmp/build/80754af9/nbclient_1614364831625/work
nbconvert @ file:///tmp/build/80754af9/nbconvert_1624479060632/work
nbformat @ file:///tmp/build/80754af9/nbformat_1617383369282/work
nest-asyncio @ file:///tmp/build/80754af9/nest-asyncio_1613680548246/work
netCDF4 @ file:///tmp/build/80754af9/netcdf4_1624383287733/work
notebook @ file:///tmp/build/80754af9/notebook_1621528346532/work
numpy @ file:///tmp/build/80754af9/numpy_and_numpy_base_1620830962040/work
numpyro @ file:///home/conda/feedstock_root/build_artifacts/numpyro_1615923251938/work
olefile==0.46
opt-einsum @ file:///tmp/build/80754af9/opt_einsum_1621500238896/work
packaging @ file:///tmp/build/80754af9/packaging_1611952188834/work
pandas==1.2.5
pandocfilters @ file:///tmp/build/80754af9/pandocfilters_1605120460739/work
parso @ file:///tmp/build/80754af9/parso_1617223946239/work
pathspec==0.8.1
patsy==0.5.1
pexpect @ file:///tmp/build/80754af9/pexpect_1605563209008/work
pickleshare @ file:///tmp/build/80754af9/pickleshare_1606932040724/work
Pillow @ file:///tmp/build/80754af9/pillow_1617383569452/work
pluggy @ file:///tmp/build/80754af9/pluggy_1615976321666/work
poetry-semver==0.1.0
poetry2conda==0.3.0
prometheus-client @ file:///tmp/build/80754af9/prometheus_client_1623189609245/work
prompt-toolkit @ file:///tmp/build/80754af9/prompt-toolkit_1616415428029/work
psutil @ file:///tmp/build/80754af9/psutil_1612298023621/work
ptyprocess @ file:///tmp/build/80754af9/ptyprocess_1609355006118/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl
py @ file:///tmp/build/80754af9/py_1607971587848/work
pycparser @ file:///tmp/build/80754af9/pycparser_1594388511720/work
Pygments @ file:///tmp/build/80754af9/pygments_1621606182707/work
pyparsing @ file:///home/linux1/recipes/ci/pyparsing_1610983426697/work
pyrsistent @ file:///tmp/build/80754af9/pyrsistent_1600141720057/work
pytest==5.4.3
python-dateutil @ file:///home/ktietz/src/ci/python-dateutil_1611928101742/work
pytz @ file:///tmp/build/80754af9/pytz_1612215392582/work
PyYAML==5.4.1
pyzmq==20.0.0
qtconsole @ file:///tmp/build/80754af9/qtconsole_1623278325812/work
QtPy==1.9.0
questionary==1.9.0
regex==2021.4.4
scipy @ file:///tmp/build/80754af9/scipy_1618855647378/work
seaborn @ file:///tmp/build/80754af9/seaborn_1608578541026/work
Send2Trash @ file:///tmp/build/80754af9/send2trash_1607525499227/work
sip==4.19.13
six @ file:///tmp/build/80754af9/six_1623709665295/work
statsmodels @ file:///tmp/build/80754af9/statsmodels_1614023746358/work
tabulate==0.8.9
termcolor==1.1.0
terminado==0.9.4
testpath @ file:///tmp/build/80754af9/testpath_1624638946665/work
toml @ file:///tmp/build/80754af9/toml_1616166611790/work
tomlkit==0.7.2
tornado @ file:///tmp/build/80754af9/tornado_1606942300299/work
tqdm @ file:///tmp/build/80754af9/tqdm_1623943362311/work
traitlets @ file:///home/ktietz/src/ci/traitlets_1611929699868/work
typing-extensions @ file:///home/ktietz/src/ci_mi/typing_extensions_1612808209620/work
wcwidth @ file:///tmp/build/80754af9/wcwidth_1593447189090/work
webencodings==0.5.1
widgetsnbextension==3.5.1
xarray @ file:///tmp/build/80754af9/xarray_1620826949436/work
zipp @ file:///tmp/build/80754af9/zipp_1615904174917/work
```

## Code Snippet
```python
Python 3.8.10 (default, Jun  4 2021, 15:09:15) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: J = 8

In [3]: y = np.array([28.0, 8.0, -3.0, 7.0, -1.0, 1.0, 18.0, 12.0])

In [4]: sigma = np.array([15.0, 10.0, 16.0, 11.0, 9.0, 11.0, 10.0, 18.0])

In [5]: import numpyro

In [6]: import numpyro.distributions as dist

In [7]: def eight_schools(J, sigma, y=None):
   ...:     mu = numpyro.sample('mu', dist.Normal(0, 5))
   ...:     tau = numpyro.sample('tau', dist.HalfCauchy(5))
   ...:     with numpyro.plate('J', J):
   ...:         theta = numpyro.sample('theta', dist.Normal(mu, tau))
   ...:         numpyro.sample('obs', dist.Normal(theta, sigma), obs=y)
   ...: 

In [8]: from jax import random

In [9]: from numpyro.infer import MCMC, NUTS

In [10]: nuts_kernel = NUTS(eight_schools)

In [11]: mcmc = MCMC(nuts_kernel, num_warmup=500, num_samples=1000)
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)

In [12]: rng_key = random.PRNGKey(0)

In [13]: mcmc.run(rng_key, J, sigma, y=y, extra_fields=('potential_energy',))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-5d1f91ec73c6> in <module>
----> 1 mcmc.run(rng_key, J, sigma, y=y, extra_fields=('potential_energy',))

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    496         map_args = (rng_key, init_state, init_params)
    497         if self.num_chains == 1:
--> 498             states_flat, last_state = partial_map_fn(map_args)
    499             states = tree_map(lambda x: x[jnp.newaxis, ...], states_flat)
    500         else:

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    331         rng_key, init_state, init_params = init
    332         if init_state is None:
--> 333             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
    334                                            model_args=args, model_kwargs=kwargs)
    335         sample_fn, postprocess_fn = self._get_cached_fns()

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    503         )
    504         if rng_key.ndim == 1:
--> 505             init_state = hmc_init_fn(init_params, rng_key)
    506         else:
    507             # XXX it is safe to run hmc_init_fn under vmap despite that hmc_init_fn changes some

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/numpyro/infer/hmc.py in <lambda>(init_params, rng_key)
    486                              ' `potential_fn`.')
    487 
--> 488         hmc_init_fn = lambda init_params, rng_key: self._init_fn(  # noqa: E731
    489             init_params,
    490             num_warmup=num_warmup,

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/numpyro/infer/hmc.py in init_kernel(init_params, num_warmup, step_size, inverse_mass_matrix, adapt_step_size, adapt_mass_matrix, dense_mass, target_accept_prob, trajectory_length, max_tree_depth, find_heuristic_step_size, forward_mode_differentiation, model_args, model_kwargs, rng_key)
    209         """
    210         step_size = lax.convert_element_type(step_size, jnp.result_type(float))
--> 211         trajectory_length = lax.convert_element_type(trajectory_length, jnp.result_type(float))
    212         nonlocal wa_update, max_treedepth, vv_update, wa_steps, forward_mode_ad
    213         forward_mode_ad = forward_mode_differentiation

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/_src/lax/lax.py in convert_element_type(operand, new_dtype)
    423   if hasattr(operand, '__jax_array__'):
    424     operand = operand.__jax_array__()
--> 425   return _convert_element_type(operand, new_dtype, weak_type=False)
    426 
    427 def _convert_element_type(operand: Array, new_dtype: Optional[DType] = None,

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/_src/lax/lax.py in _convert_element_type(operand, new_dtype, weak_type)
    452     return operand
    453   else:
--> 454     return convert_element_type_p.bind(operand, new_dtype=new_dtype,
    455                                        weak_type=new_weak_type)
    456 

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/core.py in bind(self, *args, **params)
    262         args, used_axis_names(self, params) if self._dispatch_on_params else None)
    263     tracers = map(top_trace.full_raise, args)
--> 264     out = top_trace.process_primitive(self, tracers, params)
    265     return map(full_lower, out) if self.multiple_results else full_lower(out)
    266 

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/core.py in process_primitive(self, primitive, tracers, params)
    601 
    602   def process_primitive(self, primitive, tracers, params):
--> 603     return primitive.impl(*tracers, **params)
    604 
    605   def process_call(self, primitive, f, tracers, params):

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/interpreters/xla.py in apply_primitive(prim, *args, **params)
    246 def apply_primitive(prim, *args, **params):
    247   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
--> 248   compiled_fun = xla_primitive_callable(prim, *unsafe_map(arg_spec, args), **params)
    249   return compiled_fun(*args)
    250 

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/interpreters/xla.py in arg_spec(x)
    238 
    239 def arg_spec(x: Any) -> ArgSpec:
--> 240   aval = abstractify(x)
    241   try:
    242     return aval, x._device

~/anaconda3/envs/numpyro_play/lib/python3.8/site-packages/jax/interpreters/xla.py in abstractify(x)
    184   if hasattr(x, '__jax_array__'):
    185     return abstractify(x.__jax_array__())
--> 186   raise TypeError(f"Argument '{x}' of type '{type(x)}' is not a valid JAX type")
    187 
    188 def _make_abstract_python_scalar(typ, val):

TypeError: Argument 'None' of type '<class 'NoneType'>' is not a valid JAX type
```

Below is the environment where I don't see any issue, using the same system and same code snippet.

```bash
 conda list | grep jax
jax                       0.2.10                   pypi_0    pypi
jaxlib                    0.1.62                   pypi_0    pypi

# packages in environment at /home/benda/anaconda3/envs/numpyro_test:
numpyro                   0.6.0                    pypi_0    pypi

python --version
Python 3.7.10

```

**Comment by fehiepsi** [comment]
Hi @bdatko, could you install numpyro with `pip install -q numpyro@git+https://github.com/pyro-ppl/numpyro`? We will release a new version soon.

**Comment by bdatko** [comment]
Yup, that fixed the error. Thanks @fehiepsi

**Comment by bmorris3** [comment]
This error also resolved for me after building numpyro from source.

**Comment by fehiepsi** [comment]
We have just released numpyro 0.7. Please check it out. :)
